packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -146,169 +150,144 @@ impl PostgresDatabase {
-        )
-        .bind(created_by)
-        .fetch_one(&self.pool)
-        .await
-        .report()
-        .change_context(InsertionError)
-        .attach_lazy(|| data_type.clone())?;
+    ) -> Result<Qualified<T>, InsertionError>
+    where
+        T: DatabaseType + Serialize + Send + Sync,
+    {
+        let uri = database_type.uri();
 
-        Ok(())
+        if self
+            .contains_base_uri(uri.base_uri())
+            .await
+            .change_context(InsertionError)?
+        {
+            return Err(Report::new(BaseUriAlreadyExists)
+                .attach_printable(uri.base_uri().clone())
+                .change_context(InsertionError));
+        }
+
+        self.insert_base_uri(uri.base_uri()).await?;
+
+        let version_id = self.insert_uri(uri).await?;
+
+        self.insert_with_id(version_id, &database_type, created_by)
+            .await?;
+
+        Ok(Qualified::new(version_id, database_type, created_by))
     }
 
-    /// Inserts a [`PropertyType`] identified by [`VersionId`], and associated with an
-    /// [`AccountId`], into the database.
+    /// Updates the specified [`DatabaseType`].
+    ///
+    /// This first checks, if the [`BaseUri`] of the type does exist. It will then creates a new
+    /// [`VersionId`] from the contained [`VersionedUri`] and inserts the type.
     ///
     /// # Errors
     ///
-    /// - [`DatastoreError`], if inserting failed.
-    async fn insert_property_type(
-        &mut self,
-        version_id: VersionId,
-        property_type: &PropertyType,
-        created_by: AccountId,
-    ) -> Result<(), InsertionError> {
-        sqlx::query_as::<_, (Uuid,)>(
-            r#"
-            INSERT INTO property_types (
-                version_id,
-                schema,
-                created_by
-            ) 
-            VALUES ($1, $2, $3)
-            RETURNING version_id;
-            "#,
-        )
-        .bind(version_id)
-        .bind(
-            serde_json::to_value(property_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize property type: {err}")),
-        )
-        .bind(created_by)
-        .fetch_one(&self.pool)
-        .await
-        .report()
-        .change_context(InsertionError)
-        .attach_lazy(|| property_type.clone())?;
+    /// - If the [`BaseUri`] does not already exist
+    ///
+    /// [`BaseUri`]: crate::types::BaseUri
+    async fn update<T>(
+        &self,
+        database_type: T,
+        updated_by: AccountId,
+    ) -> Result<Qualified<T>, UpdateError>
+    where
+        T: DatabaseType + Serialize + Send + Sync,
+    {
+        let uri = database_type.uri();
 
-        Ok(())
+        if !self
+            .contains_base_uri(uri.base_uri())
+            .await
+            .change_context(UpdateError)?
+        {
+            return Err(Report::new(BaseUriDoesNotExist)
+                .attach_printable(uri.base_uri().clone())
+                .change_context(UpdateError));
```

Comment:
General `error-stack` comment, not for this PR:

I think we should either get the calling function to do this (change_context), or we should write a macro (if we can) to do this. Because this is a lot of boilerplate and it's showing a problem with our approach in my eyes. I don't think we should have to duplicate the exact same change_context on every single return value in the function.